optests improvements based on torchvision usage on nms

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108929

- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over
- Support dynamic shapes by default for make_fx `tracing_mode="fake"` without symbolifying everything else

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
optests improvements based on torchvision usage on nms

- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "optests improvements based on torchvision usage on nms"


- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "optests improvements based on torchvision usage on nms"


- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "optests improvements based on torchvision usage on nms"


- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "optests improvements based on torchvision usage on nms"


- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

zou3519 commented on torch/_prims_common/__init__.py:
```diff
@@ -77,11 +77,16 @@
 
     for x, y in zip(a, b):
+        if allow_rhs_unbacked:
+            # TODO: We should check that the symbols are consistent
+            # with each other
+            if isinstance(y, torch.SymInt):
+                continue
```
What's the idea here?

ezyang replied:
If I run a data-dependent op, I will end up with an unbacked symint on the fake tensor. I cannot directly compare the unbacked symint with the real symint. So I just say "well, whatever, if the fake value is symint, let's just assume it's fine." It's definitely possible to do a more strict check but I was lazy.

zou3519 commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -352,7 +351,7 @@ def numpy_nms_abstract(boxes, scores, iou_threshold):
     ctx = torch._custom_op.impl.get_ctx()
     i0 = ctx.create_unbacked_symint()
-    result = boxes.new_empty([i0, 4])
+    result = boxes.new_empty([i0], dtype=torch.int64)
```
Test failures look real. It looks like make_fx(tracing_mode='fake') now doesn't raise DynamicOutputShapeException? 
<img width="905" alt="image" src="https://github.com/pytorch/pytorch/assets/5652049/ad6f001a-f902-495c-956a-b35071e529c4">

https://github.com/pytorch/pytorch/blob/9a492fc27f3f1f0f71ed8b24704430a87a99cec9/test/test_custom_ops.py#L1459-L1467

zou3519 reviewed (commented):
Code looks good, test failures might be real (did we change intentionally change the behavior of make_fx(tracing_mode=fake) on dynamic output shape operations?)

ezyang commented on torch/_subclasses/fake_utils.py:
```diff
@@ -79,9 +81,12 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         ):
             try:
-                with FakeTensorMode() as fake_mode:
+                # enable_python_dispatcher() here
```
whoops sorry, is a TODO

ezyang commented:
> Code looks good, test failures might be real (did we change intentionally change the behavior of make_fx(tracing_mode=fake) on dynamic output shape operations?)

So, I am finding that I have to fix latent bugs now that I have enabled data-dependent output shapes.

The main issue is that `dynamic_only=True` doesn't work anymore: previously, we would just test if an exception was raised, but now no exception is raised so the test as written doesn't actually make sense. Along the way, it seems the meta for NMS in testing isn't right lol.

ezyang committed:
Update on "optests improvements based on torchvision usage on nms"


- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over
- Support dynamic shapes by default for make_fx `tracing_mode="fake"` without symbolifying everything else

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

zou3519 reviewed (approved):
SGTM

ezyang committed:
Update on "optests improvements based on torchvision usage on nms"


- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over
- Support dynamic shapes by default for make_fx `tracing_mode="fake"` without symbolifying everything else

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -162,7 +162,7 @@ def numpy_nonzero_impl(x):
     ctx = torch._custom_op.impl.get_ctx()
     i0 = ctx.create_unbacked_symint()
-    shape = [x.dim(), i0]
+    shape = [i0, x.dim()]
```
the tuple axis (which is being added) is dim length, so if you insert it at dim=1 it is at the end. You could resolve this either way but this was the minimum change.

ezyang committed:
Update on "optests improvements based on torchvision usage on nms"


- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over
- Support dynamic shapes by default for make_fx `tracing_mode="fake"` without symbolifying everything else

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "optests improvements based on torchvision usage on nms"


- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over
- Support dynamic shapes by default for make_fx `tracing_mode="fake"` without symbolifying everything else

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -f "ci failures look spurious"

ezyang committed:
Update on "optests improvements based on torchvision usage on nms"


- Update cross-ref FakeMode test to use ShapeEnv.  Dynamic ops can now
  return an unbacked SymInt.  We always accept this as equal to whatever
  the real value was.
- Relax test so it works on all classes, not just unittest.TestCase
- Properly wrap the original method, so things like
  pytree.mark.parametrize are carried over
- Support dynamic shapes by default for make_fx `tracing_mode="fake"` without symbolifying everything else

Fixes https://github.com/pytorch/pytorch/issues/108927

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -r

ezyang commented:
@pytorchbot merge -f "known master breakage"
